File: public/donations.py
"""
Public API endpoints for donations.

This module provides public routes for creating donations.
All endpoints are accessible without authentication.
"""

from flask import request, jsonify
from forms import DonationForm
from models import db, Donation, Drive
from utils import update_setting
from . import public_bp

# -------------------------------
# API: Donations
# -------------------------------

@public_bp.route('/donations', methods=['POST'])
def create_donation():
    """
    Public endpoint to create a new donation.
    Expects JSON with donor name, amount, optional message and anonymity flag.
    Returns the created donation ID and a success message.
    No authentication required.
    """
    data = request.get_json()
    form = DonationForm(data=data, meta={'csrf': False})
    if not form.validate():
        return jsonify({'errors': form.errors}), 400
    
    drive_id = data.get('drive_id')
    if drive_id is not None:
        drive = Drive.query.get(drive_id)
        if not drive:
            return jsonify({'error': 'Указанный сбор не существует'}), 400
        if drive.status != 'активен':
            return jsonify({'error': 'Выбранный сбор не активен'}), 400
    
    donation = Donation(
        donor_name=form.name.data,
        amount=form.amount.data,
        message=form.message.data,
        is_anonymous=form.is_anonymous.data,
        status='ожидает',
        drive_id=drive_id
    )
    db.session.add(donation)
    db.session.commit()
    
    # Update stats
    sum_donation = db.session.query(db.func.sum(Donation.amount)).scalar()
    update_setting('sum_donation', str(sum_donation))
    
    return jsonify({'message': 'Пожертвование принято', 'id': donation.id}), 201

File: static/js/src/index/donations.js
import { reactive, ref } from 'vue';
import { fetchStats } from './stats.js';
import { showNotification } from './notification.js';

export const donationForm = reactive({
    name: '', amount: 1000, message: '', is_anonymous: false
});
export const donationLoading = ref(false);
export const showDonationModal = ref(false);
export const selectedDriveId = ref(null);

export async function submitDonation() {
    donationLoading.value = true;
    try {
        const response = await fetch('/api/public/donations', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                name: donationForm.is_anonymous ? 'Аноним' : donationForm.name,
                amount: donationForm.amount,
                message: donationForm.message,
                is_anonymous: donationForm.is_anonymous,
                drive_id: selectedDriveId.value
            })
        });
        if (response.ok) {
            await fetchStats();
            showNotification('Спасибо за ваше пожертвование!', 'success');
            donationForm.name = ''; donationForm.amount = 1000; donationForm.message = '';
            donationForm.is_anonymous = false;
            showDonationModal.value = false;
        } else {
            const err = await response.json();
            showNotification(err.errors || 'Ошибка при отправке', 'error');
        }
    } catch (e) {
        showNotification('Ошибка сети', 'error');
    } finally {
        donationLoading.value = false;
    }
}

export function showDonationForDrive(driveId) {
    showDonationModal.value = true;
}